### PR TITLE
PDB fixes and testing

### PR DESCRIPTION
**Pull request by tmat**
**Customer scenario**

1) Customer debugs an iterator or async method in a VS .NET Core app. Hoisted local variables are not displayed correctly since the VB compiler didn't emit hoisted scope info to Portable PDBs.

2) Customer builds a C# app that contains partial async void method. The compiler crashes.

**Bugs this fixes:**

Fixes https://github.com/dotnet/roslyn/issues/17934: Async partial method crashes emission of portable debug info
Fixes https://github.com/dotnet/roslyn/issues/18029: Avoid emitting debug information for methods without sequence points
Fixes https://github.com/dotnet/roslyn/issues/8473: Emit State Machine Hoisted Local Scopes information as defined by the Portable PDB specification

**Workarounds, if any**

No workarounds.

**Risk**

Small. 

**Performance impact**

None

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

**How was the bug found?**

Unfinished Portable PDB support in VB and crash reported by customer.

### CODE REVIEW

**Comment by tmat** [comment]
@dotnet/roslyn-compiler Please review.
cston @ivanbasov for changes in EE.

File: src/Compilers/CSharp/Test/Emit/PDB/PDBAsyncTests.cs [hunk 1]
```diff
@@ -1199,7 +1200,7 @@ class C
           <namespace usingCount=""1"" />
         </using>
         <dynamicLocals>
-          <bucket flagCount=""1"" flags=""1"" slotId=""1"" localName=""d"" />
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/PDB/PDBAsyncTests.cs]: flagCount is redundant, it's always the length of flags.

**Comment by tmat** [comment]
test windows_debug_unit32_prtest  please

**Comment by tmat** [comment]
test windows_debug_unit32_prtest please

**Comment by tmat** [comment]
test windows_debug_unit32_prtest  please
test windows_debug_unit64_prtest  please
test windows_release_unit32_prtest  please

**Comment by tmat** [comment]
test windows_debug_unit32_prtest please

**Comment by tmat** [comment]
test windows_debug_unit64_prtest please

**Comment by tmat** [comment]
test windows_release_unit32_prtest please

**Comment by tmat** [comment]
@dotnet/roslyn-compiler Ping

File: src/Compilers/Core/Portable/CodeGen/MethodBody.cs [hunk 2]
```diff
@@ -85,12 +80,21 @@ internal sealed class MethodBody : Cci.IMethodBody
             _stateMachineHoistedLocalScopes = stateMachineHoistedLocalScopes;
             _stateMachineHoistedLocalSlots = stateMachineHoistedLocalSlots;
             _stateMachineAwaiterSlots = stateMachineAwaiterSlots;
+            _stateMachineMoveNextDebugInfoOpt = stateMachineMoveNextDebugInfoOpt;
             _dynamicAnalysisDataOpt = dynamicAnalysisDataOpt;
+            _sequencePoints = GetSequencePoints(sequencePoints, debugDocumentProvider);
         }
 
-        void Cci.IMethodBody.Dispatch(Cci.MetadataVisitor visitor)
+        private ImmutableArray<Cci.SequencePoint> GetSequencePoints(SequencePointList sequencePoints, DebugDocumentProvider debugDocumentProvider)
```
Comment (cston) [inline src/Compilers/Core/Portable/CodeGen/MethodBody.cs]: `static`?

File: src/Compilers/Core/Portable/Emit/AsyncMoveNextBodyDebugInfo.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        public readonly ImmutableArray<int> ResumeOffsets;
+
+        public AsyncMoveNextBodyDebugInfo(
+            Cci.IMethodDefinition kickoffMethod,
+            int catchHandlerOffset,
+            ImmutableArray<int> yieldOffsets,
+            ImmutableArray<int> resumeOffsets) 
+            : base(kickoffMethod)
+        {
+            Debug.Assert(!yieldOffsets.IsDefault && !resumeOffsets.IsDefault && yieldOffsets.Length == resumeOffsets.Length);
```
Comment (cston) [inline src/Compilers/Core/Portable/Emit/AsyncMoveNextBodyDebugInfo.cs]: Consider splitting into three separate asserts.

File: src/Compilers/Core/Portable/Emit/StateMachineMoveNextDebugInfo.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.Emit
+{
+    /// <summary>
+    /// Information associated with methdo body of a state machine MoveNext method.
```
Comment (cston) [inline src/Compilers/Core/Portable/Emit/StateMachineMoveNextDebugInfo.cs]: Typo: `method body`

File: src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb [hunk 5]
```diff
@@ -1501,44 +1501,64 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
                 Dim stateMachineMethod = TryCast(method, SynthesizedStateMachineMethod)
+                If stateMachineMethod IsNot Nothing AndAlso method.Name = WellKnownMemberNames.MoveNextMethodName Then
+
+                    kickoffMethod = stateMachineMethod.StateMachineType.KickoffMethod
+                    Debug.Assert(kickoffMethod IsNot Nothing)
+
+                    isAsyncStateMachine = kickoffMethod.IsAsync
+
+                    ' Async Sub may be partial. Debug info needs to be associated with the emitted definition, 
+                    ' but the kickoff method Is the method implementation (the part with body).
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb]: Typo: `is`

File: src/Test/Utilities/Portable/Pdb/PdbValidation.cs [hunk 6]
```diff
@@ -220,29 +207,55 @@ private static void RemoveEmptyMethods(XElement pdb)
...
+        private static void RemovePortableSpecificElements(XElement expectedNativePdb)
+        {
+            var elementsToRemove = from e in expectedNativePdb.DescendantsAndSelf()
+                                   where e.Attributes().Any(a => a.Name.LocalName == "format" && a.Value == "portable")
+                                   select e;
+
+            foreach (var e in elementsToRemove.ToArray())
+            {
+                e.Remove();
+            }
```
Comment (cston) [inline src/Test/Utilities/Portable/Pdb/PdbValidation.cs]: The same `foreach (var e in elementsToRemove.ToArray()) { ... }` appears several times. Consider extracting a `RemoveElements(IEnumerable<XElement>)` helper method.

**Comment by cston** [comment]
Minor comments only. LGTM.

File: src/Compilers/Core/Portable/CodeGen/MethodBody.cs (hunk 2 above)
Comment (tmat) [inline src/Compilers/Core/Portable/CodeGen/MethodBody.cs]: Fixed

File: src/Compilers/Core/Portable/Emit/AsyncMoveNextBodyDebugInfo.cs (hunk 3 above)
Comment (tmat) [inline src/Compilers/Core/Portable/Emit/AsyncMoveNextBodyDebugInfo.cs]: Done

File: src/Compilers/Core/Portable/Emit/StateMachineMoveNextDebugInfo.cs (hunk 4 above)
Comment (tmat) [inline src/Compilers/Core/Portable/Emit/StateMachineMoveNextDebugInfo.cs]: Fixed

File: src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb (hunk 5 above)
Comment (tmat) [inline src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb]: Fixed

File: src/Test/Utilities/Portable/Pdb/PdbValidation.cs (hunk 6 above)
Comment (tmat) [inline src/Test/Utilities/Portable/Pdb/PdbValidation.cs]: Fixed.

**Comment by tmat** [comment]
test windows_eta_open_prtest please

**Comment by tmat** [comment]
@dotnet/roslyn-compiler Need 2nd review.

**Comment by tmat** [comment]
windows_eta_open_prtest failure is unrelated

**Comment by tmat** [comment]
windows_eta_open_prtest failure is unrelated.

**Comment by tmat** [comment]
@MattGertz for ask mode approval.

File: src/Compilers/Core/Portable/PEWriter/CustomDebugInfoWriter.cs [hunk 7]
```diff
@@ -83,14 +83,6 @@ public byte[] SerializeMethodDebugInfo(EmitContext context, IMethodBody methodBo
             var pooledBuilder = PooledBlobBuilder.GetInstance();
             var encoder = new CustomDebugInfoEncoder(pooledBuilder);
 
-            // NOTE: This is an attempt to match Dev10's apparent behavior.  For iterator methods (i.e. the method
```
Comment (agocke) [inline src/Compilers/Core/Portable/PEWriter/CustomDebugInfoWriter.cs]: Why is this comment no longer relevant?

Comment (tmat) [inline src/Compilers/Core/Portable/PEWriter/CustomDebugInfoWriter.cs]: The "unusual" behavior in ISymUnmanagedWriter has been fixed a while ago.  I don't think we need to document what Dev10 does anymore either.

**Pull request merged**